Verif adj ctrl gen

mjlosch opened the pull request:
## What changes does this PR introduce?
remove ECCO_CTRL_DEPRECATED from some verification experiments to use the nice GENARR/GENTIM2D variant
- [x] global_ocean.cs32x15
- [x] isomip
- [x] obcs_ctrl
- [x] offline_exf_seaice
- [ ] tutorial_global_oce_optim

## What is the current behaviour? 
most verification experiments use the flag ECCO_CTRL_DEPRECATED, see issue #401 
 
## What is the new behaviour 
more verification experiments use GENARR/GENTIM2D control variables, also requires fixing some bugs, e.g. in `thsice` code, make `ctrlprec` a runtime parameters, etc.

## Does this PR introduce a breaking change? 
hopefully not

## Suggested addition to `tag-index`
o replace cpp flag global_ocean.cs32x15 in global_ocean.cs32x15, isomip, obcs_ctrl, and offline_exf_seaice    verification experiments by the much nicer generic control variables GENARR/GENTIM2D variant. This required a number of changes to the general code to make the generic control variables work in a non-ecco context:
  - pkg/thsice: add AUTODIFF_OPTIONS.h and COST_OPTIONS.h header files to numerous routines
  - pkg/seaice: mask tendencies of HEFF/AREA/HSNOW in seaice_growth_adx.F to reproduce cost function values
  - pkg/shelfice: enable GENARR/GENTIM2D methods
  - pkg/obcs: move reading control variable out of S/R obcs_prescribe_read into S/R obcs_calc to be independent of parameter values unrelated to the adjoint
  - pkg/ctrl: make generic control variable work without pkg/cal (works only for time independent control variables), but issue a warning, bonus: make ctrlprec a runtime parameter that defaults to 64
  - pkg/ecco: reorder CPP flags in ecco_cost_weights.F and remove unused variable 'wbaro', move calling S/R cost_obcs out of ECCO_CTRL_DEPRECATED code block in S/R ecco_cost_driver, reorganise CPP flags in cost_averagesinit.F
  - model/src/the_main_loop.F: fix tape initialisation, bonus: suppress a benign warning
  - for each of the verification experiments remove ECCO_CPPOPTIONS.h and replace by CTRL_OPTIONS.h, CTRL_SIZE.h, sometimes COST/ECCO_OPTIONS.h, EXF_OPTIONS.h,  adjust data.ctrl, data.grdchk, data.ecco
  - for global_ocean.cs32x15 compile both variants by keeping global_ocean.cs32x15 in CTRL_OPTIONS.h and use runtime parameter ctrlUseGen to switch between methods (in global_ocean.cs32x15.thsice). This involves an extra (benign) recomputation warning this verification experiment because of an if statement with ctrlUseGen in ctrl_init_variables.F

mjlosch commented:
@jm-c I am stuck at debugging this, because I don't know exactly where these adxx_files are written. Do you (or anyone else) have a clue?

mjlosch commented:
OK, found the problem with the adxx_files: the GEN*_CONTROL requires a weights file, which needs to have the correct floating point precision. Unfortunately, one cannot omit this file. I personally think that the weights should default to 1, if nothing is specified and I might add this to this PR, if no-one objects.

mjlosch commented:
@jm-c the last two commits are for global_ocean.cs32x15. Maybe you can have quick look and then we should chat about this, because I don't know if this is going in the right direction.

mjlosch reviewed (commented):
This:
  Y Y Y Y  4>16< 9 pass  offline_exf_seaice  (e=0, w=37)
is related to a masking problem. Suddenly HEFF has nonzero values over land. Adding a mask to the cost function quick-fixes this, but does not get to the root of the problem.

mjlosch commented:
As this is work in progress, I just undid something that @jm-c did 9 years ago and with that fixed the masking problem in offline_exf_seaice. The global_ocean.cs32x15 is not affected by this. I have to say, that this turns out to be far more painful than expected. Not sure how many more I still want to (o:

mjlosch commented:
Instead of 51724b9, the masking of `HEFF`/`HSNOW`/`AREA` should probably better be done directly in `seaice_growth_adx.F`.

mjlosch commented:
I think that I am more or less done with this if we want to leave `tutorial_global_oce_optim` as it is for now. There are two problems remaining:

- [ ] global_ocean.cs32x15 does not give exactly the same numbers. This is due to activated code in autodiff_init_varia.F, per flag `ALLOW_AUTODIFF_INIT_OLD` (not sure why this is needed here). With this flag and `ECCO_CTRL_DEPRECATED`, this code resets `EmPmR=0` within initialise_varia.F and the experiment starts with no FWflux in the first time step (`EmPmR` stats = 0 in monitor output). When `ECCO_CTRL_DEPRECATED` is undefined, this code is not active anymore and we start with non-zero `EmPmR`, as it should be. I think this is a bug in the experiment (and the old code) and we should update the output_adm file for this experiment. I can check in an AUTODIFF_OPTIONS.h file that has this option undefined.
- [ ] FD agreement in obcs_ctrl is only 10 in my tests, but cost and AD is 16 digits. Not clear to me why, probably because of different precision of weights file `errorTtot.err`, that is required with the new version in order to keep the cost and AD correct.

timothyas reviewed (commented):
Thanks for adding all of this @mjlosch. I just gave the code a once over and it generally looks pretty detailed and good to go from what I was able to see. To reply to your comments:
- I would also be in favor of a default "one" value for weights if a file is not provided. I basically end up carrying around a "ones.data" file everywhere for this reason, which would be nice to not have to remember.
- I agree with the changes to the "exact numbers" in global_ocean.cs32x15. I don't understand why the old code would set `EmPmR=0`, unless it is a remnant from code that tried to avoid making adjustments that were too large in the first few time steps. I hear about that a lot, but I have no idea. Either way, I agree with this change.

> FD agreement in obcs_ctrl is only 10 in my tests, but cost and AD is 16 digits. Not clear to me why, probably because of different precision of weights file errorTtot.err, that is required with the new version in order to keep the cost and AD correct.

- Is the old test single precision and now it's double?

jm-c commented:
@mjlosch I am starting to look at this PR. The first thing I was going to do was to merge in master (it's easier to compare with
latest results). The other thing is that you could fill up the section "addition to tag-index suggestion", including the list of experiments
that have been updated, it would be useful (78 modified files, a little bit of help there cannot hurt).

jm-c commented:
@mjlosch looking at global_ocean.cs32x15, the 3 secondary tests fail because of missing data.cal ; seems something is not right.
Also, I would prefer not to change global_ocean.cs32x15/input_ad/data.pkg (I found the additional commented line not really helpful there).

mjlosch commented:
@jm-c 
thanks for having a look at this. I realise that this has become a much more elaborate effort than initially intended (as you say: 78 files changed, but many of them are either related to the reference experiments or are benign, i.e. just additional options files included or similar). I had put off adding tag-index suggestion because it was not clear at the time how many experiment I would manage to convert. Now this is backfiring, because I need to remember myself (via electronic archeology), what I actually did. Will do so later today. Same with global_ocean.cs32x15, should be easy to fix, I probably just forgot to add a file or so (although it did run on two different platforms when I left this in December).

mjlosch commented:
@jm-c
Sorry about the confusion in global_ocean.cs32x15. I put back the data.cal files, don't know why I removed them in the first place. Probably because there were so many identical namelist files in these input_ad* directories that I thought one copy of each would be enough. If fact, that's what I did to many of the other namelist files and now link them from input_ad if possible (see my comment on Dec17). If that's not a good policy, then we can revert that part easily. 

I also reverted input_ad/data.pkg. I think temporarily I had data.cal turned on in this experiment as well (seemed to be required for the GENTIM2D variables), but then I was able to turn it off again, so this is a left over and not intentional.

The main remaining issues that I see are in my comment of 22 Dec 2020. And we need to agree on what to do with tutorial_global_oce_optim. I would prefer to leave it alone for now and not convert to generic control variables.

jm-c commented:
@mjlosch OK, things are better now, and the list you added for "tag-index addition" helps. I will take a look to confirm the reason
why global_ocean.cs32x15 output are changing (your Dec 22 comment) but obcs_ctrl does not look that bad (will check also).
Since there are already many changes in this PR, I suggest to leave tutorial_global_oce_optim unchanged for now, could be done in an other PR.
And one general comment about duplicated parameter files in secondary input(_af).* dir: For forward test (i.e., input.*) I like to keep multiple version of "eedata" to enable a direct and easy comparison between the multi-threaded (eedata.mth) and single thread version (and in this case, we really needs eedata.mth in each secondary input dir to get the secondary tested by testreport -mth). Also, since data.pkg is short anyway but useful in getting a quick overview of which pkg is used, we could be flexible for this one and leave some duplicate copies.
For the other parameter files, I agree it's better to remove duplicate copies (as you did here).

mjlosch commented:
@jm-c 
You can recover the old output of global_ocean.cs32x15 by removing the ECCO_CTRL_DEPRECATED condition in autodiff_init_varia.F 
The question is why this code is there in the first place and maybe we just want to ignore the ALLOW_AUTODIFF_INIT_OLD flag and update the reference instead.

jm-c commented:
Regarding obcs_ctrl, I am getting a very good agreement (16 digits for all FWD monitor, 16 digits for Cost & AD-gradient) so it's great. The Fwd-gradients are the same up to 9.digits which is very good (since it's computed from differences of very similar numbers).

jm-c commented:
@mjlosch Regarding global_ocean.cs32x15: I am checking how to reproduce the new results (from this PR) using the unmodified/original code. But it's relatively clear to me that this ALLOW_AUTODIFF_INIT_OLD is just a way to put back
an old bug. Now it it works well, assuming I just have to change ECCO_CPPOPTIONS.h, we could agree on:
1) make a new PR just to update this experiment (ECCO_CPPOPTIONS + 4 AD output + 4 TLM).
2) go back to this PR, merge in master after (1) is merged, and will be easier to finish.
Because ECCO_CPPOPTIONS is gone in this current PR (and are not planning to update ref. output), there should not be
merging conflicts (or easy to fix). And this way a "bug fix" will not be buried into a full set of experiment config updates.
The alternative would be to put back the bug here (so that results don't change) and fix it later. The third alternative where both are in the same current PR is less appealing (already many changes here).

jm-c commented:
@mjlosch I confirm that switching off ALLOW_AUTODIFF_INIT_OLD in current (unmodified) global_ocean.cs32x15 set-up
gives exactly the same results ad in this PR.

mjlosch commented:
@jm-c 
good suggestion about global_ocean.cs32x15. A separate PR that prepends this one is the cleanest solution. Can I ask you to do that since you have an easier access to the reference machines than I do? Or do you want me to start it (make the change in ECCO_CPPOPTIONS.h)?

jm-c commented:
I will start this small, new PR today.

jm-c commented:
OK, I made a new (small) PR #436 just to fix AD experiment global_ocean.cs32x15. Will come back to this PR after my new one is merged in.

mjlosch commented:
@jm-c great!
Do you think it is useful to add your commit message about ECCO_CTRL_DEPRECATED to AUTODIFF_OPTIONS.h?

mjlosch commented:
On my linux box (which is new to me and I don't have to much experience how close it is to the reference machine) I am getting now with gfortran [gcc version 9.3.0 (Ubuntu 9.3.0-17ubuntu1~20.04)]:
```
default    10
G D M    C  A  F
e p a R  o  d  D
n n k u  s  G  G
2 d e n  t  r  r

Y Y Y Y 16>16<16 pass  global_ocean.cs32x15  (e=0, w=43)
Y Y Y Y 16>16<16 pass  global_ocean.cs32x15.seaice
Y Y Y Y  9>12< 3 pass  global_ocean.cs32x15.seaice_dynmix
Y Y Y Y 16>16< 9 pass  global_ocean.cs32x15.thsice
Y Y Y Y 16>16< 9 pass  isomip  (e=0, w=4)
Y Y Y Y 16>16<16 pass  isomip.htd
Y Y Y Y 16>14< 8 pass  obcs_ctrl  (e=0, w=0)
Y Y Y Y 16>16<16 pass  offline_exf_seaice  (e=0, w=33)
Y Y Y Y 16>16<16 pass  offline_exf_seaice.thsice
Start time:  Sat 27 Feb 2021 12:35:34 PM CET
End time:    Sat 27 Feb 2021 12:57:30 PM CET
```
Is that expected (global_ocean.cs32x15.seaice_dynmix)?

jm-c commented:
> On my linux box (which is new to me and I don't have to much experience how close it is to the reference machine) I am getting now with gfortran [gcc version 9.3.0 (Ubuntu 9.3.0-17ubuntu1~20.04)]:
> 
> ```
> default    10
> G D M    C  A  F
> e p a R  o  d  D
> n n k u  s  G  G
> 2 d e n  t  r  r
> 
> Y Y Y Y 16>16<16 pass  global_ocean.cs32x15  (e=0, w=43)
> Y Y Y Y 16>16<16 pass  global_ocean.cs32x15.seaice
> Y Y Y Y  9>12< 3 pass  global_ocean.cs32x15.seaice_dynmix
> Y Y Y Y 16>16< 9 pass  global_ocean.cs32x15.thsice
> Y Y Y Y 16>16< 9 pass  isomip  (e=0, w=4)
> Y Y Y Y 16>16<16 pass  isomip.htd
> Y Y Y Y 16>14< 8 pass  obcs_ctrl  (e=0, w=0)
> Y Y Y Y 16>16<16 pass  offline_exf_seaice  (e=0, w=33)
> Y Y Y Y 16>16<16 pass  offline_exf_seaice.thsice
> Start time:  Sat 27 Feb 2021 12:35:34 PM CET
> End time:    Sat 27 Feb 2021 12:57:30 PM CET
> ```
> 
> Is that expected (global_ocean.cs32x15.seaice_dynmix)?

I can diff the STDOUT and see no difference between this PR and (now updated) master branch for this experiment (seaice_dynmix). I guess this one is more sensitive to compiler version.
But regarding global_ocean.cs32x15.thsice, there are some differences in AD monitor for fields ad_exf_adfu & ad_exf_adfv
that are larger than machine truncation. Not sure why. But nothing shows up in Cost/AD-Grad/FWD-Grad.

jm-c commented:
I think it would be nice, at some point, to get a more precise idea about which header file should be used to #define CPP options that have been sitting in the "one file for all" ECCO_CCPOPTIONS.h ; for instance, ECCO_CTRL_DEPRECATED is listed in
ECCO_OPTIONS.h but will not work for many src files that use it (because ECCO_OPTIONS.h is not included in these src files),
including autodif_init_varia.F. May be a better place for this one would be CTRL_OPTIONS.h ?

jm-c commented:
@mjlosch I am going to change some ALLOW_AUTODIFF_TAMC (when not related to TAF storage directives) to simply ALLOW_AUTODIFF in the pkg/thsice S/R you updated. It's somehow related to the previous changes (that were needed to have ALLOW_AUTODIFF_TAMC defined) so it fits well here.

mjlosch commented:
> I think it would be nice, at some point, to get a more precise idea about which header file should be used to #define CPP options that have been sitting in the "one file for all" ECCO_CCPOPTIONS.h ; for instance, ECCO_CTRL_DEPRECATED is listed in ECCO_OPTIONS.h but will not work for many src files that use it (because ECCO_OPTIONS.h is not included in these src files), including autodif_init_varia.F. May be a better place for this one would be CTRL_OPTIONS.h ?

The problem is that this flag really does not conform to our "standards", because it affects to two packages (ECCO and CTRL) and it is needed in both and unfortunately, not all files that include ECCO_OPTIONS.h also include CTRL_OPTIONS.h and vice versa. Probably, it is best to *not* have this flag (and corresponding options) show up at all in any header file other than ECCO_CPPOPTIONS.h? I had kept the commented flag so far, so that we can easily switch back and forth between old (deprecated) and generic cost/ctrl.
The only alternative I see is to put it into CPP_OPTIONS.h (because it is not specific to just one pkg).

mjlosch commented:
> I can diff the STDOUT and see no difference between this PR and (now updated) master branch for this experiment (seaice_dynmix). I guess this one is more sensitive to compiler version. 

OK, I can reproduce that, but there's an intriguing difference still (old < > new):
```
< (PID.TID 0000.0001) ctrl-wet 1:    nvarlength =       239366
---
> (PID.TID 0000.0001) ctrl-wet 1:    nvarlength =       239768
[...]
< (PID.TID 0000.0001) ctrl_init: control vector length:          239366
---
> (PID.TID 0000.0001) ctrl_init: control vector length:          239768
```
and more differences for 'ctrl_init_wet.F'. Not sure where that comes from. 

> But regarding global_ocean.cs32x15.thsice, there are some differences in AD monitor for fields ad_exf_adfu & ad_exf_adfv that are larger than machine truncation. Not sure why. But nothing shows up in Cost/AD-Grad/FWD-Grad.

It is reassuring that it only happens with `thsice` (because that's probably not used that much with AD), but it's discomforting that the test numbers do not pick it up. I guess I don't understand, how the reference value for AD-Grad is determined, just from the gradient check? In that case we may want to rethink ...

jm-c commented:
@mjlosh regarding where  IECCO_CTRL_DEPRECATED should be defined, took a look at pkg/ecco, seems to me that we only need to add CTRL_OPTIONS.h in 12 src files ; so not too bad. 
Along the way, I saw things that are not right, e.g., ALLOW_AUTODIFF_TAMC is used in ecco_cost_driver.F but AUTODIFF_OPTIONS.h is missing (and this might have an effect on global_oce_cs32 & global_oce_llc90, will check).

And regarding global_ocean.cs32x15.thsice, could we try - just for the purpose of understanding the changes in AD monitor - to
change the perturbation ctrl variable that are sampled in GrdCheck to test this exf_adfu, exf_adfv ?

mjlosch commented:
> And regarding global_ocean.cs32x15.thsice, could we try - just for the purpose of understanding the changes in AD monitor - to change the perturbation ctrl variable that are sampled in GrdCheck to test this exf_adfu, exf_adfv ?

yes, we can do that. For that we need to perturb xx_fu (grdchkvarindex = 303) or xx_fv (grdchkvarindex = 304), I think, but I can't do that today. 
It is very likely that something went wrong in the thsice pkg (because the exf_adfu are OK with the seaice experiments), either after splitting up ECCO_CPPOPTIONS.h (3898b4e) or when I tried to make it work with the generic controls (0152fae). Since I have done that in separate steps, it should be possible to find out, which step has lead to the problem, again something for tomorrow.

mjlosch commented:
@jm-c 
I found two issues so far
1. the generic controls always use "'c'" (hardwired) as the grid type (variable `ncvargrd`) so that the masks are sometimes wrong if, e.g. the forcing is defined on u or v points (such as `fu` and `fv`). This not a problem, just for counting and packing the control vector and it explains the different control vector lengths. It does not fix the differences in ad_exf_adfu/v.
2. unrelated: there are two fields `tmpfld2d` and `tmpfld3d` defined in `ctrl.h`. They used to be in a common block (commented out), but I don't understand why they are still there, instead of being defined locally in the few (I found seven) routines where they are used.

should we fix that here?

jm-c commented:
@mjlosch these 2 arrays that are in header file but not in common block should be declared locally, I agree. But it's not directly related to ECCO_CTRL_DEPRECATED so it could be done in an other PR (cleaning stuff).

mjlosch commented:
@jm-c after merging your stuff, I will push some changes about the masking. Unfortunately this involves a little more work, because of the shelfice package, where we have yet another mask.

About the problem with adfu/v: 
1. Not sure if that helps, but I see a difference between tr_run.thsice and e.g. tr_run.seaice: at the end of exf_getforcing.F we have: 
     ```
      if ( .not. useseaice) then
        call exf_adjoint_snapshots( 3,mytime,myiter,mythid )
      endif
     ``` 
      Instead, this is called from the end of seaice_model. In tr_run.thsice we do not use seaice so that the admonitor with iwhen=3 is called in different place in these two experiments. Maybe that can give us a clue.
2. Also I have the impression that the original "ALLOW_TAUU/V0_CONTROL" does not correspond exactly to the new xx_gentim2d_file="xx_fu/v" (although I don't see why not), because even the finite difference gradient is not the same (or even similar). Unfortunately I cannot test the ALLOW_H/SFLUX0_CONTROL, because that's not implemented with pkg/grdchk. Here I am not sure if I use pkg/grdchk correctly.

mjlosch commented:
@jm-c 
The good news is, that the fields `fu_ad/fv_ad` are actually not different except for a missing or incorrect exchange somewhere (e.g. the are different along face edges and this shows up in admonitory output). Not yet sure where this happens. It turns out that the generic controls and ECCO_CTRL_DEPRECATED can be compiled in the same executable and then you can switch between the two version with the runtime parameter `ctrlUseGen`. I find that with `ctrlUseGen=.TRUE.` I get the problem we are observing and with `ctrlUseGen=.FALSE.` the problem goes away, i.e. this difference along the face boundaries vanishes. It's no clear to me, which of the two is actually correct, either.

The bad news is, that the gradients that are written to the disk (`adxx_tauu/v` in the old case and `adxx_fu/v` with `gentim2d`) differ approximately by a factor of 4 (but not uniform, large fluctuations). I have no idea where these differences are produced. `adxx_tauu/v` are written directly in `ctrl_map_forcing_ad`, so there's not that much that can happen, while in the generic control case `fu_ad/fv_ad` are copied to `xx_gentim2d_ad`, which is then passed to `xx_gentim2d_loc_ad` in `ctrl_map_gentim2d_ad`, which is then passed to `ctrl_gen_get_ad`, where more stuff is happening. So maybe we cannot expect these files to be the same? I think they should be, because nominally both are `fu/fv_ad`.

mjlosch commented:
@jm-c
I think that the difference in the variables `fu/v_ad` is due to a (benign?) problem in the old code. See the attached figure, where you can clearly see the line in the centre of the domain in the output of `ADJtauy` (which is `fv_ad`), `sNy = 16`, the top figures looks much better:
![ADJtauy4](https://user-images.githubusercontent.com/30285667/109972991-2508cf00-7cf8-11eb-91fe-a2beb3b2a80f.png)

I am not sure how much time I want to spend debugging old (DEPRECATED) code, so I would argue that we should not worry about the different monitor output.

I am worried about the differences between `adxx_fu/fv` and `adxx_taux/tauy`, which lead to the different values in the gradient checks (if one tests these fields). It's entirely unclear to me, where these large difference come from. Here's an example, that has been created after compiling the DEPRECATED code together with the generic control variables (move their definitions out of the ifdef/else in CTRL_OPTIONS.h) and running with `ctrlUseGen=.TRUE.`, both fields are from the same run:
![adxx_fu](https://user-images.githubusercontent.com/30285667/109974980-68643d00-7cfa-11eb-9a56-5594e8e10887.png)

mjlosch commented:
I dug a little deeper and the field `xx_gentim2d_ad_loc`, that is passed on to `ctrl_get_gen_ad`, still has the correct values (e.g., compared to `adxx_taux` and variable `fu_ad`), but something is going on in `ctrl_get_gen_ad` that changes this. I guess it's because we have constant forcing, but then map this to the field "1" and "0" or so? So yes, it may have to do with a time integration effect, but for the other old code there's also time integration, so I don't know ...

More ctrl-blues: I really like the generic control variables much better than the old stuff (everything is so much easier), but they assume the same mask (always maskC) for every variable. That's mostly OK, but there are few fields, where it's not (e.g. fu,fv,u/vstress on C-grid, shifwflx, ...). I am not sure if that's a problem now, or if we should just gloss over this here and have a different PR, where these things are cleaned up? What do you think.

jm-c commented:
@mjlosch  I also see more output files like: adxx_fu.tmp and adxx_fu.effective (don't know what they are).
And regardng fixing some mask in some ctrl, would definitively be better in an other PR.

timothyas commented:
Hey @mjlosch, regarding the ctrl-blues and masking, I did this in one of one of my other branches where I was working with control variables with shelfice. I created functions to get the appropriate mask for each variable see e.g.

https://github.com/timothyas/MITgcm/blob/shi_trans_coeff_ctrl/pkg/ctrl/ctrl_get_mask.F

And whenever a mask is needed there are calls to this, something like:

```
call ctrl_get_mask2d(xx_genarr2d_file(iarr), mask2D, myThid)
```

or similar for 3D. I'd be happy to either cherry pick and commit those changes, or commit it with that branch, which adds the shelfice exchange coefficients as control variables within the generic ctrl framework. It will need some cleanup before making a PR out of it though ...

mjlosch commented:
@timothyas
cool, that looks exactly like the contribution we need here. But as @jm-c points out, it may be the wrong PR for this.
In this PR, somewhat off-topic but necessary to make `isomip` work, I added `xx_shifwflx` to the generic control framework, so I'd appreciate, if you could check that part of this PR (since you obviously have done something similar). I also added new runtime flag that let's you manually prescribe the correct flag. I like your solution a little better, but maybe more expensive?

@jm-c and @timothyas I feel like this PR is too much of mess to add anything new. I would hold of any changes for now, hope that it gets merged soon. I don't think that we need to sort out the differences between adxx_taux of the DEPRECATED method and adxx_fu of the generic control parameters here, because that does not affect the verification experiments.
I suggest to have 2 new PR, one that address #437 and one that "cleans up" the generic control variables (masking and shelfice, etc.) and the ctrl package in general (I have a few small things that will help in the future). Or even separate the second into two (one reserved to shelfice). What do you think?

jm-c commented:
@mjlosch Sounds a good plan. Also in pkg/ctrl cleaning, I would suggest to move to pkg/ctrl the customized version of
ctrl_map_ini_genarr.F that is sitting in: verification/global_oce_biogeo_bling/code_ad

timothyas commented:
Sure @mjlosch I can take a look at the isomip/gentim2d parts. In general, I agree with what you've suggested, and reserving any additional features to future PRs - this seems like it is already record breaking in terms of number of line or file changes! And I agree with the two new PRs, and I think the shelfice doesn't need to be separate from the rest of the "clean up", but perhaps we can cross the bridge when we get there ... 


--- 

(Perhaps you're right @mjlosch, it is a bit more expensive. Maybe there's a better way to do it which we can work out in the forthcoming PR, but in general I like the idea of reducing the number of inputs the user must specify when they dramatically impact the fundamental behavior of the controls, and so prefer to have the masks hard coded based on the control variable rather than specifying the `xx_gentim2d_grid`).

mjlosch commented:
@jm-c @timothyas 
great, so there's already a long list of things to do in future PRs. I will start an issue where we can collect them before actually starting the PR?
But to move forward, what are the things, we need to do for this PR (since this should be finished before doning anything new)? To my mind, this is left:

- [ ] Do we need to sort out the difference between `adxx_fu` and variable `fu_ad`? It means going into `ctrl_get_gen.F`. I would postpone that to the separate PR.

mjlosch commented:
@jm-c I think I finally know where the difference between adxx_fu and adxx_taux comes from. The accumulation of the adjoint variables over time seems to be different between the versions of the verification experiment global_ocean.cs32x15.
1. In the DEPRECATED code in global_ocean.cs32x15 with ALLOW_TAUU0_CONTROL, we read xx_taux (= 0) from file only at the beginning of the integration (because it does not change in time) and also only add it to `fu` for `myIter==nIter0` (in ctrl_map_forcing). Consequently, `fu_ad` is accumulated backwards over time and then saved to adxx_taux in ctrl_map_forcing_ad at myIter=nIter0 only. The result is the gradient of the cost function to the stress forcing a nIter0 (the beginning of the integration). This makes sense for initial conditions, for a (constant) forcing field it is not clear to me, how useful this is.
2. With `ctrlUseGen`, `xx_fu` (=xx_gentim2d_ad=0) is also only read at the beginning of the integration (because it too is constant in time, with the corresponding xx_gen_period = 0), but then it is added to fu at **every** time step. That makes sense to me, because if you want to perturb your forcing by a constant field, then you have to add it at all times (and not just at the beginning). Consequently, `fu_ad` is accumulated backwards over time and then saved to `xx_gentim2d_ad = xx_gentim2d_ad + fu_ad` each time step. This looks like it is accumulated a second time, but that's of course correct, because we add `xx_gentim2d` each time step. In ctrl_map_gentim2d_ad, `xx_gentim2d_ad` is copied to `xx_gentim2d_loc_ad` and afterwards zeroed out so that the second accumulation actually wasn't one. `xx_gentim2d_loc_ad` is then passed to ctrl_get_gen_ad each tilmestep, and there it is then distributed to the two time levels (although in this case only level 0 is filled, here it is now really accumulated for the second time as it should be), but only written to a file for nIter0. The result is a gradient of the cost function to stress forcing at each time step, which makes far more sense to me than the gradient just wrt the initial forcing at `nIter0`.

So in summary, the two versions use different control variables and with the generic controls it's not possible to reproduce control variables like ALLOW_TAUU0_CONTROL. When you remove the `IF ( myIter .EQ. nIter0 )` in ctrl_map_forcing.F then `adxx_fu` and `adxx_taux` are the same, because then `adxx_taux` is accumulated (each time step) by the inner workings of `s/r adactive_file_read`. 

By switching this experiment to the generic control variables, we loose testing the special case of forcing control variables that give gradients only wrt to the first time step. I think that this is a loss that we can live with (as this type of gradient is physically difficult to interpret).

mjlosch commented:
So I think, we are done here. One thing one may want to do is to change the CTRL_OPTIONS.h file(s) to compile both generic control variables and DEPRECATED code, so that one can switch between these if desired and even keep the original result for global_ocean.cs32x15.thsice.

jm-c commented:
@mjlosch This is good news !  Everything is now sorted out.
Now regarding switching back some exp to compile DEPRECATED code, we could but we don't really need this now since there are still a decent number of exp. that use this DEPRECATED code. Or may be we just do it in 1 of the 4 experiment ?
And somehow related: I was going to update some comments around ECCO_CTRL_DEPRECATED in few code_ad dir
(I think it could be useful since the name might let users think it belongs to ECCO_OPTIONS.h).

timothyas commented on pkg/shelfice/shelfice_thermodynamics.F:
```diff
@@ -216,30 +222,44 @@ SUBROUTINE SHELFICE_THERMODYNAMICS(
 #if (defined SHI_ALLOW_GAMMAFRICT && defined ALLOW_DIAGNOSTICS)
           uStarDiag             (I,J,bi,bj) = 0. _d 0
 #endif /* SHI_ALLOW_GAMMAFRICT and ALLOW_DIAGNOSTICS */
+#if (defined ALLOW_SHIFWFLX_CONTROL || defined ALLOW_GENTIM2D_CONTROL)
```
Is there a reason to keep the `ALLOW_SHIFWFLX_CONTROL` flag around, other than to support old code? I would suggest to replace it simply with `ALLOW_GENTIM2D_CONTROL`, as is done with all other gentim2d variables.

It also exists in shelfice_:
- ad_check_lev1/2/3/4_dir.h
- cost_shifwflx.F
- cost_final.F
- init_fixed

My vote is for removing it and adding something to shelfice_check.F to warn the user, either at compile time or run time, about the replacement. That is, unless there's good reason to keep it and I'm just missing it. I'd be happy to make those changes @mjlosch if that's decided upon.

timothyas commented on pkg/shelfice/shelfice_thermodynamics.F:
```diff
@@ -216,30 +222,44 @@ SUBROUTINE SHELFICE_THERMODYNAMICS(
+#endif /* ALLOW_GENTIM2D_CONTROL */
+#if (defined ECCO_CTRL_DEPRECATED && defined ALLOW_SHIFWFLX_CONTROL)
+C     unfortunately, we need this CPP flag here, because otherwise
+C     none of these fields are defined
+      ELSEIF ( useCTRL ) THEN
+        CALL CTRL_GET_GEN (
      &     xx_shifwflx_file, xx_shifwflxstartdate, xx_shifwflxperiod,
      &     maskSHI, xx_shifwflx_loc, xx_shifwflx0, xx_shifwflx1,
      &     xx_shifwflx_dummy,
      &     xx_shifwflx_remo_intercept, xx_shifwflx_remo_slope,
      &     wshifwflx,
      &     myTime, myIter, myThid )
-#endif
-#endif /* ALLOW_SHIFWFLX_CONTROL */
+#endif /* ECCO_CTRL_DEPRECATED and ALLOW_SHIFWFLX_CONTROL */
+      ENDIF
+#endif /* ALLOW_CTRL */
```
I guess your comment here answers my earlier question. I still vote for doing away with `ALLOW_SHIFWFLX_CONTROL` altogether. Although perhaps that can be saved for #440

mjlosch replied:
I think getting rid of this flag is a good suggestion, but let's save this for #440 because this PR was really only about turning off the DEPRECATED code in some more of the verification experiment.

timothyas replied:
That sounds good to me

jm-c commented:
@mjlosch What about compiling DEPRECATED code in obcs_ctrl but still using the generic ctrl there ? it would provide an easy way to get back the original results. Also, since it's the only experiment with OBC control (and one a the few using pkg/ecco), it would make sure the DEPRECATED code is still compiling.

timothyas commented:
@mjlosch @jm-c I ran all of the adjoint verification experiments and the only notable issue seemed to be `obcs_ctrl`. It turns out that ecco_cost_final needs to have the variables `no_obcs*` pulled out of the DEPRECATED macro, that is https://github.com/MITgcm/MITgcm/blob/43034d44525429c1d24df94da91c7de32917e2ab/pkg/ecco/ecco_cost_final.F#L369-L372

needs to be moved out of the ECCO_CTRL_DEPRECATED (for some reason, `f_obcs*` already are, but these didn't make the cut somehow). I'd be happy to commit that, or it's pretty straightforward if someone else wants to. With this fix, I get exact same agreement (cost, adj, and FD) to master with that experiment.

Other than that I only had the following differences to an old "master" run that I had lying around from my mdsio PR. It ends up with 1 or 2 digits of difference - I'm not sure how much it matters, or how much you guys have already covered this earlier, but I'll just report these differences below.

This PR:
```
Y Y Y Y 16>14< 9 pass  global_ocean.cs32x15  (e=0, w=43)
Y Y Y Y  8>11< 2 pass  global_ocean.cs32x15.seaice_dynmix
Y Y Y Y 16>16< 9 pass  global_ocean.cs32x15.thsice
```

"Master" (commit 66fe84e)
```
Y Y Y Y 16>15< 9 pass  global_ocean.cs32x15  (e=0, w=43)
Y Y Y Y 16>15< 2 pass  global_ocean.cs32x15.seaice_dynmix
Y Y Y Y 16>14<10 pass  global_ocean.cs32x15.thsice
```

mjlosch commented:
@jm-c 
I guess, with my suggestion, I intended to have an example where both ECCO_CTRL_DEPRECATED and some ALLOW_GEN*_CONTORL is defined. I would still use the generic controls in that case (`ctrlUseGen=.TRUE.`), but one could also think of having some "sub-"experiments use `ctrlUseGen=.TRUE.` and one `ctrlUseGen=.FALSE.`. In the case of `global_ocean.cs32x15`, this could be `global_ocean.cs32x15.thsice`, because then nothing in the output changes. With `obcs_ctrl` this wouldn't work and doesn't make sense because the obcs-ctrl-varaibles are not part of the generic controls.

jm-c commented:
@mjlosch I agree, and my suggestion about obcs_ctrl does not make much sense.

jm-c commented:
@mjlosch Thanks for adding these useful comments in this "data.ctrl" !

jm-c reviewed (commented):
This is great work that is very helpful in switching and testing generic-controls.
I will go through part of the changes later (but there are many changes).
In the mean time, I tried to run TLM testreport but got some issue with lab_sea (does not finish due to MITgcm error) and the 4 global_ocean.cs32x15 tlm results are not good. But this needs to be confirmed.

jm-c commented:
> In the mean time, I tried to run TLM testreport but got some issue with lab_sea (does not finish due to MITgcm error)

No problem with TLM for lab_sea (did not clean-up well lab_sea run dir before running the first time).

mjlosch commented:
@jm-c I forgot to post this comment before pushing my new data.ctrl. I guess it's not necessary anymore, but here's what I intended to do (and you change check, if I succeeded (o:):

I agree with your comment about data.ctrl. I'll update data.ctrl in the following way (because I think this it is instructive to see how to switch between the two methods):
1. I'll reinsert the namelist parameters for the old method (only the ones that are actually used), but these are are the default values that are set in ctrl_readparms.F anyway
2. I'll keep the generic control variable namelist parameters.
3. I'll add an explanation of all of this at they top of the file

mjlosch commented:
@jm-c 
I, too, get this for the TLM:
```
Y Y Y Y  4> 9<16 FAIL  global_ocean.cs32x15
Y Y Y Y  3> 4<16 FAIL  global_ocean.cs32x15.seaice
Y Y Y Y  3> 4< 3 FAIL  global_ocean.cs32x15.seaice_dynmix
Y Y Y Y 16>14< 9 pass  global_ocean.cs32x15.thsice
```
This is another "gem" of the ctrl package. Somewhere ctrl_map_ini.F we have this:
```
#ifndef DISABLE_CTRL_THETA_LIMIT
                if(theta(i,j,k,bi,bj).lt.-2.0)
     &               theta(i,j,k,bi,bj)= -2.0
#endif
```
The initial theta has a few values below -2.0, so this is corrected here in the old version of the control variables. In the generic control variables, this is done with the `xx_genarr3d_bounds` vector, but because of an incorrect `ctrl_bounds.flow`, the ctrl_bounds2d/3d function calls are not included in the tangent linear model. After correcting the flow directives, the forward model is exact again. I'll push a fix soon. (This was really hard to find!!)

jm-c commented:
@mjlosch I can imagine it was "hard to find" since I don't even see how the fix works, but the good thing is that I now get all the same TLM gradient check as before on my laptop, even for global_ocean.cs32x15:
Y Y Y Y 16>16<16 pass  global_ocean.cs32x15
Y Y Y Y 16>16<16 pass  global_ocean.cs32x15.seaice
Y Y Y Y  9>12< 3 pass  global_ocean.cs32x15.seaice_dynmix
Y Y Y Y 16>16<16 pass  global_ocean.cs32x15.thsice
(and checked that seaice_dynmix output is unchanged).

jm-c commented:
I run the adjoint experiments from verification_other and no changes (2 less recomputation warning) so everything is good.
@mjlosch  I made a minor adjustment to a new warning message (I thought could be misinterpreted) but you can change this
if you feel like. Otherwise, from my side, we could merge this PR soon.

mjlosch commented:
@jm-c 
great! I'll (squash-) merge this later today (after adjusting the tag-index). With the squash-merge we loose all the painful history of this (which may be useful for future efforts like this), but I'll try to keep as many useful comments as possible.
